platform/api/oci/types/instance_test.go

```diff
@@ -0,0 +1,399 @@
+// Copyright 2017 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestLaunchInstanceInput(t *testing.T) {
+	type in struct {
+		input LaunchInstanceInput
+	}
+	type out struct {
+		jsonStr string
+	}
+
+	tests := []struct {
+		in   in
+		out  out
+		name string
+	}{
+		{
+			in: in{input: LaunchInstanceInput{
+				AvailabilityDomain: "Uocm:PHX-AD-1",
+				CompartmentID:      "ocid1.compartment.oc1..aaaaaaaayzfqeibduyox6iib3olcmdar3ugly4fmameq4h7lcdlihrvur7xq",
+				CreateVNICDetails: &VNICDetails{
+					AssignPublicIP:      BoolToPtr(false),
+					DisplayName:         StrToPtr("example-vnic-details"),
+					HostnameLabel:       StrToPtr("bminstance-1"),
+					PrivateIP:           StrToPtr("10.0.3.3"),
+					SkipSourceDestCheck: BoolToPtr(true),
+					SubnetID:            "example-subnet-id",
+				},
+				DisplayName:      StrToPtr("example-instance"),
+				HostnameLabel:    StrToPtr("bminstance-1"),
+				ImageID:          "ocid1.image...",
+				IPXEScript:       StrToPtr("example-ipxe-script"),
+				Metadata:         &map[string]string{"foo": "bar"},
+				Shape:            "VM.Standard1.1",
+				SubnetID:         StrToPtr("example-subnet-id"),
+				ExtendedMetadata: RawMessageToPtr(json.RawMessage(`{"extended":{"metadata":"example"}}`)),
+			}},
+			out:  out{jsonStr: `{"availabilityDomain":"Uocm:PHX-AD-1","compartmentId":"ocid1.compartment.oc1..aaaaaaaayzfqeibduyox6iib3olcmdar3ugly4fmameq4h7lcdlihrvur7xq","createVnicDetails":{"assignPublicIp":false,"displayName":"example-vnic-details","hostnameLabel":"bminstance-1","privateIp":"10.0.3.3","skipSourceDestCheck":true,"subnetId":"example-subnet-id"},"displayName":"example-instance","hostnameLabel":"bminstance-1","imageId":"ocid1.image...","ipxeScript":"example-ipxe-script","metadata":{"foo":"bar"},"shape":"VM.Standard1.1","subnetId":"example-subnet-id","extendedMetadata":{"extended":{"metadata":"example"}}}`},
```

Comment:
This style of testing the marshalled string doesn't seem great.

Notably, things like non-deterministic ordering of maps could easily bite these tests.

Reply:
Would it be preferred to `Unmarshal` the output back into an object and compare the objects or build regular expression based asserts on the JSON output?